syntax = "proto2";

package settings;

import "point3d.proto";

message PrinterSettings {
    required string name = 1 [default = "New Printer"];
    required int32 guid = 2;
    optional string description = 3;

    message MovementSettings {
        optional float retract_length = 1 [default = 2.0];
        optional float retract_speed = 2 [default = 6000];
        optional float deretract_speed = 3 [default = 3000];
        optional float movement_speed = 4 [default = 6000];
        optional float z_hop = 5 [default = 0.5];
        optional float z_hop_speed = 6 [default = 6000];
    }

    optional MovementSettings movement_settings = 4;

    message PrinterBounds {
        optional float min_x = 1;
        optional float max_x = 2;
        optional float min_y = 3;
        optional float max_y = 4;
        optional float min_z = 5;
        optional float max_z = 6;
    }
    optional PrinterBounds printer_bounds = 5;

    optional Point3D origin = 6;

    optional string snapshot_command = 7 [default = "snap"];
    optional float printer_position_confirmation_tolerance = 8 [default = 0.001];
    optional bool auto_detect_position = 9 [default = true];
    optional bool abort_out_of_bounds = 10 [default = true];
    optional bool override_octoprint_print_volume = 11 [default = false];

    repeated string auto_position_detection_commands = 12;
    optional float priming_height = 13 [default = 0.75];

    enum AxisMode {
        REQUIRE_EXPLICIT = 1;
        RELATIVE = 2;
        ABSOLUTE = 3;
    }
    optional AxisMode e_axis_default_mode = 14 [default = REQUIRE_EXPLICIT];
    optional AxisMode xyz_axes_default_mode = 15 [default = REQUIRE_EXPLICIT];

    enum TrueFalseUseOctoprint {
        TRUE = 1;
        FALSE = 2;
        USE_OCTOPRINT = 3;
    }
    optional TrueFalseUseOctoprint g90_influences_extruder = 16 [default = USE_OCTOPRINT];

    optional string units_default = 17 [default="millimeters"];

    enum SpeedUnit {
        MM_MIN = 1;
        IN_MIN = 2;
    }
    optional SpeedUnit axis_speed_display_units = 18 [default=MM_MIN];

    optional bool default_firmware_retractions = 19 [default=false];
    optional bool default_firmware_retractions_zhop = 20 [default=false];
}
